pkg/server/policy/policy.go

```diff
@@ -0,0 +1,170 @@
+package policy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/storage"
+	"github.com/open-policy-agent/opa/storage/inmem"
+)
+
+const (
+	allowKey             = "allow"
+	allowIfAdminKey      = "allow_if_admin"
+	allowIfDownstreamKey = "allow_if_downstream"
+	allowIfAgentKey      = "allow_if_agent"
+	allowIfLocalKey      = "allow_if_local"
+)
+
+// Engine drives policy management.
+type Engine struct {
+	rego rego.PartialResult
+}
+
+type EngineConfig struct {
+	FileProvider *FileProviderConfig `hcl:"file_provider"`
+}
+
+type FileProviderConfig struct {
+	PolicyPath      string `hcl:"policy_path"`
+	PermissionsPath string `hcl:"permissions_path"`
+}
+
+// Input represents context associated with an access request.
+type Input struct {
+	// Caller is the authenticated identity of the actor making a request.
+	Caller string `json:"caller"`
+
+	// FullMethod is the fully-qualified name of the proto rpc service method.
+	FullMethod string `json:"full_method"`
+
+	// Req represents data received from the request body. It MUST be a
+	// protobuf request object with fields that are serializable as JSON,
+	// since they will be used in policy definitions.
+	Req interface{} `json:"req"`
+}
+
+type Result struct {
+	Allow             bool `json:"allow"`
+	AllowIfAdmin      bool `json:"allow_if_admin"`
+	AllowIfLocal      bool `json:"allow_if_local"`
+	AllowIfDownstream bool `json:"allow_if_downstream"`
+	AllowIfAgent      bool `json:"allow_if_agent"`
+}
+
+// NewEngineOrDefault returns a new policy engine. Or if no
+// config is provided, provides the default policy
+func NewEngineOrDefault(ctx context.Context, cfg *EngineConfig) (*Engine, error) {
+	if cfg == nil {
+		return DefaultAuthPolicy(ctx)
+	}
+	return NewEngine(ctx, cfg)
+}
+
+// NewEngine returns a new policy engine. Or nil if no
+// config is provided.
+func NewEngine(ctx context.Context, cfg *EngineConfig) (*Engine, error) {
```

Comment:
This function is not exercised in the tests. I think that would be good to have it covered.

Reply:
Yup - created tests for `NewEngineFromConfigOrDefault`, and made this function private, since it is just a helper. Should encapsulate all the edge cases for `NewEngine`.